api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -827,21 +815,185 @@ private String generateRandomEightCharacterQualifier() {
+      throw new FailedPreconditionException(
+          "The workspace CDR version does not have whole genome data");
+    }
+
+    if (dataSetExportRequest.getKernelType().equals(KernelTypeEnum.R)) {
+      return generateGenomicsAnalysisComment_R();
+    }
+
+    switch (dataSetExportRequest.getGenomicsAnalysisTool()) {
+      case HAIL:
+        return generateHailCode(qualifier, dataSetExportRequest);
+      case PLINK:
+        return generatePlinkCode(qualifier, dataSetExportRequest);
+      case NONE:
+        return generateDownloadVcfCode(qualifier, dataSetExportRequest);
+      default:
+        throw new BadRequestException("Invalid Genomics Analysis Tool");
+    }
+  }
+
+  private Optional<String> getExtractionDirectory(Long datasetId) {
+    try {
+      return Optional.of(
+              submissionDao
+                  .findMostRecentValidExtractionByDataset(dataSetDao.findById(datasetId).get())
+                  .get()
+                  .getOutputDir()
+                  .replaceFirst("/$", "")) // Drop trailing slash if exists
+          .filter(dir -> !dir.isEmpty());
+    } catch (NoSuchElementException e) {
+      return Optional.empty();
+    }
+  }
+
+  private String generateVcfDirEnvName(String qualifier) {
+    return "DATASET_" + qualifier + "_VCF_DIR";
+  }
+
+  private static final String MISSING_EXTRACTION_DIR_PLACEHOLDER =
+      "\"GCS_VCF_DIRECTORY_GOES_HERE\"";
+
+  private String generateExtractionDirCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    String extractionDir =
+        getExtractionDirectory(dataSetExportRequest.getDataSetRequest().getDataSetId())
+            .orElse(MISSING_EXTRACTION_DIR_PLACEHOLDER);
+
+    String noExtractionDirComment =
+        MISSING_EXTRACTION_DIR_PLACEHOLDER.equals(extractionDir)
+            ? "# VCF files for this dataset do not exist\n"
+                + "# Run a Genomic Extraction from a Dataset to generate VCF files\n"
+            : "";
+
+    return noExtractionDirComment
+        + "%env "
+        + generateVcfDirEnvName(qualifier)
+        + "="
+        + extractionDir;
+  }
+
+  private List<String> generateHailCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    final String matrixName = "mt_" + qualifier;
+
+    return ImmutableList.of(
+        "# Note: Hail demos must be run from a \"Hail Genomics Analysis\" cloud analysis environment",
+        generateExtractionDirCode(qualifier, dataSetExportRequest),
+        "# Initialize Hail\n"
+            + "\n"
+            + "import hail as hl\n"
+            + "import os\n"
+            + "from hail.plot import show\n"
+            + "\n"
+            + "hl.init(default_reference='GRCh38')\n"
+            + "hl.plot.output_notebook()",
+        "# Create Hail Matrix table\n"
+            + "# This can take a few hours for larger (several hundred participant) cohorts\n"
```

Comment:
nit: I would not characterize this as large. I would probably just state the data point without characterizing the size, e.g. "This can take a few hours for a cohort with hundreds of participants."

Reply:
sounds good